src/hotspot/share/cds/filemap.cpp

```diff
@@ -2088,13 +2134,21 @@ void FileMapInfo::map_heap_regions_impl() {
     log_info(cds)("CDS heap data needs to be relocated lower by a further " SIZE_FORMAT
                   " bytes to " INTX_FORMAT " to be aligned with HeapRegion::GrainBytes",
                   align, delta);
-    HeapShared::init_narrow_oop_decoding(narrow_oop_base() + delta, narrow_oop_shift());
+    if (UseCompressedOops) {
+      HeapShared::init_narrow_oop_decoding(narrow_oop_base() + delta, narrow_oop_shift());
+      relocated_closed_heap_region_bottom = start_address_as_decoded_from_archive(si);
+    } else {
+      relocated_closed_heap_region_bottom = si->mapping_offset() + header()->heap_begin() + delta;
+    }
     _heap_pointers_need_patching = true;
-    relocated_closed_heap_region_bottom = start_address_as_decoded_from_archive(si);
   }
   assert(is_aligned(relocated_closed_heap_region_bottom, HeapRegion::GrainBytes),
          "must be");
 
+  if (!UseCompressedOops && delta != 0) {
+    HeapShared::set_runtime_delta(delta);
+  }
+
```

Comment:
I think the code in the previous few blocks can be simplified by adding two utility functions:

```
// The address where the bottom of this shared heap region should be mapped
// at runtime
address FileMapInfo::heap_region_runtime_start_address(FileMapRegion* spc) {
  assert(UseSharedSpaces, "runtime only");
  spc->assert_is_heap_region();
  if (UseCompressedOops) {
    return start_address_as_decoded_from_archive(si);
  } else {
     header()->heap_begin() + si->mapping_offset() + HeapShared::set_runtime_delta();
  }
}

void FileMapInfo::set_shared_heap_runtime_delta(ptrdiff_t delta) {
  if (UseCompressedOops) {
    HeapShared::init_narrow_oop_decoding(narrow_oop_base() + delta, narrow_oop_shift());
  } else {
    HeapShared::set_runtime_delta(delta);
  }
}
```

Then, many of the `if (UseCompressedOops)` checks above can be eliminated.

Reply:
For `FileMapInfo::heap_region_runtime_start_address`, I have modified it slightly to the following because
the first time it is called `_runtime_delta` hasn't been set yet.

```
address FileMapInfo::heap_region_runtime_start_address(FileMapRegion* spc, ptrdiff_t delta) {
  assert(UseSharedSpaces, "runtime only");
  spc->assert_is_heap_region();
  if (UseCompressedOops) {
    return start_address_as_decoded_from_archive(spc);
  } else {
    assert(is_aligned(spc->mapping_offset(), sizeof(HeapWord)), "must be");
    return header()->heap_begin() + spc->mapping_offset() + delta;
  }
}
```